Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp

```diff
@@ -0,0 +1,326 @@
+// Per https://www.rfc-editor.org/rfc/rfc8410
+// id-X25519    OBJECT IDENTIFIER ::= { 1 3 101 110 }
+// id-X448      OBJECT IDENTIFIER ::= { 1 3 101 111 }
+// id-Ed25519   OBJECT IDENTIFIER ::= { 1 3 101 112 }
+// id-Ed448     OBJECT IDENTIFIER ::= { 1 3 101 113 }
+// For all of the OIDs, the parameters MUST be absent.
+RefPtr<CryptoKeyOKP> CryptoKeyOKP::importSpki(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, Vector<uint8_t>&& keyData, bool extractable, CryptoKeyUsageBitmap usages)
+{
+    // FIXME: We should use the underlying crypto library to import PKCS8 OKP keys.
+
+    // Read SEQUENCE
+    size_t index = 1;
+    if (keyData.size() < index + 1)
+        return nullptr;
+
+    // Read length and SEQUENCE
+    // FIXME: Check length is 5 + 1 + 1 + 1 + keyByteSize.
+    index += bytesUsedToEncodedLength(keyData[index]) + 1;
+    if (keyData.size() < index + 1)
+        return nullptr;
+
+    // Read length
+    // FIXME: Check length is 5.
+    index += bytesUsedToEncodedLength(keyData[index]);
+    if (keyData.size() < index + 5)
+        return nullptr;
+
+    // Read OID
+    // FIXME: spec says this is 1 3 101 11X but WPT tests expect 6 3 43 101 11X.
+    if (keyData[index++] != 6 || keyData[index++] != 3 || keyData[index++] != 43 || keyData[index++] != 101)
+        return nullptr;
+
+    switch (namedCurve) {
+        case NamedCurve::X25519:
+            if (keyData[index++] != 110)
+                return nullptr;
+            break;
+        case NamedCurve::Ed25519:
+            if (keyData[index++] != 112)
+                return nullptr;
+            break;
+    };
+
+    // Read BIT STRING
+    if (keyData.size() < index + 1)
+        return nullptr;
+    if (keyData[index++] != 3)
+        return nullptr;
+
+    // Read length
+    // FIXME: Check length is keyByteSize + 1.
+    index += bytesUsedToEncodedLength(keyData[index]);
+
+    if (keyData.size() < index + 1)
+        return nullptr;
+
+    // Initial octet
+    if (!!keyData[index])
+        return nullptr;
+    ++index;
+
+    return create(identifier, namedCurve, CryptoKeyType::Public, Span<const uint8_t> { keyData.data() + index, keyData.size() - index }, extractable, usages);
+}
+
+ExceptionOr<Vector<uint8_t>> CryptoKeyOKP::exportSpki() const
+{
+    if (type() != CryptoKeyType::Public)
+        return Exception { InvalidAccessError };
+
+    size_t keySize = keySizeInBytes();
+
+    // SEQUENCE, length, SEQUENCE, length, OID, Bit String (Initial octet prepended)
+    size_t totalSize = 1 + 1 + 1 + 1 + 5 + 1 + 1 + 1 + keySize;
+    Vector<uint8_t> result;
+    result.reserveInitialCapacity(totalSize);
+    result.append(SequenceMark);
+    addEncodedASN1Length(result, totalSize - 2);
+    result.append(SequenceMark);
+    addEncodedASN1Length(result, 5);
+
+    result.append(6);
+    result.append(3);
+    result.append(43);
+    result.append(101);
+    switch (namedCurve()) {
+        case NamedCurve::X25519:
+            result.append(110);
```

Comment:
We may want to use global constants for these things. Especially given that they are used in several places (and need to match).

Reply:
done